File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/mappers/health_record_mappers/OvulationTestRecordMapper.kt
package com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers

import androidx.health.connect.client.records.OvulationTestRecord
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.OvulationTestRecordDto
import java.time.Instant
import java.time.ZoneOffset

/**
 * Converts a Health Connect [OvulationTestRecord] object to a [OvulationTestRecordDto].
 */
internal fun OvulationTestRecord.toDto(): OvulationTestRecordDto = OvulationTestRecordDto(
    id = metadata.id,
    time = time.toEpochMilli(),
    zoneOffsetSeconds = zoneOffset?.totalSeconds?.toLong(),
    metadata = metadata.toDto(),
    result = result.toOvulationTestResultDto(),
)

/**
 * Converts a [OvulationTestRecordDto] to a Health Connect [OvulationTestRecord] object.
 */
internal fun OvulationTestRecordDto.toHealthConnect(): OvulationTestRecord = OvulationTestRecord(
    time = Instant.ofEpochMilli(time),
    zoneOffset = zoneOffsetSeconds?.let { ZoneOffset.ofTotalSeconds(it.toInt()) },
    result = result.toHealthConnect(),
    metadata = metadata.toHealthConnect(id),
)

File: packages/health_connector_hk_ios/lib/src/mappers/health_record_mappers/menstruation/ovulation_test_result/ovulation_test_record_mapper.dart
import 'package:health_connector_core/health_connector_core_internal.dart';
import 'package:health_connector_hk_ios/src/mappers/health_record_mappers/health_record_id_mapper.dart';
import 'package:health_connector_hk_ios/src/mappers/health_record_mappers/menstruation/ovulation_test_result/ovulation_test_result_mapper.dart';
import 'package:health_connector_hk_ios/src/mappers/metadata_mappers/metadata_mapper.dart';
import 'package:health_connector_hk_ios/src/pigeon/health_connector_hk_ios_api.g.dart';
import 'package:meta/meta.dart' show internal;

/// Extension to convert [OvulationTestRecord] to [OvulationTestRecordDto].
@internal
extension OvulationTestRecordToDto on OvulationTestRecord {
  /// Converts [OvulationTestRecord] to [OvulationTestRecordDto].
  OvulationTestRecordDto toDto() {
    return OvulationTestRecordDto(
      id: id.value,
      metadata: metadata.toDto(),
      time: time.millisecondsSinceEpoch,
      zoneOffsetSeconds: zoneOffsetSeconds,
      result: result.toDto(),
    );
  }
}

/// Extension to convert [OvulationTestRecordDto] to [OvulationTestRecord].
@internal
extension OvulationTestRecordDtoToDomain on OvulationTestRecordDto {
  /// Converts [OvulationTestRecordDto] to [OvulationTestRecord].
  OvulationTestRecord toDomain() {
    return OvulationTestRecord.internal(
      id: id?.toDomain() ?? HealthRecordId.none,
      metadata: metadata.toDomain(),
      time: DateTime.fromMillisecondsSinceEpoch(time, isUtc: true),
      zoneOffsetSeconds: zoneOffsetSeconds,
      result: result.toDomain(),
    );
  }
}
